src/give/ContributionHistory/ContributionHistoryList.js

```
@@ -82,7 +81,14 @@ class ContributionHistoryList extends PureComponent {
       );
     }
 
-    const transactionsPerYear = map(groupBy(this.props.transactions, 'year'), (transactions, year) => ({ year, transactions }));
+    const transactionsPerYear = orderBy(
```

Could this be better handled in Highliner directly? Thoughts?

---

I think it could, but right now that might be creating more scope than we can accept (since we've already added five new stories to scope). I want the best solution for this, however I do not know the best answer as of now.

---

Will handling this client side create issues when `fetchMore` is called for additional giving history that will come down unsorted?

---

@burkeshartsis I can confirm through my local testing that fetchMore does not cause issues with this fix as seen in the GIF above. However, my testing was at a high level.

---

@IsaacHardy ordering on the server would definitely be ideal, let me have a quick look at the server

---

@Lepozepo Will the ordering need to take place here: https://github.com/NewSpring/Heighliner/blob/master/src/rock/models/finances/models/Transaction.js#L362 ?

---

@IsaacHardy it seems to be running this function to resolve transactions: https://github.com/NewSpring/Heighliner/blob/master/src/rock/models/finances/models/Transaction.js#L185 which seems to be ordering already. I'm reviewing the query, there's some nonsense going on there especially in the `.slice` part.

---

Yeah, and I think we're getting dupes too

---

@Lepozepo That would explain some odd behavior when I first started looking into this issue.

---

Ok, the silliness on the server is patched. Just need to fix tests on that branch. The duplicate keys issue is rooted in the way rerenders are being triggered. Somehow it's causing `fetchMoreResolver`'s returned function to run multiple times on the same results (which in turn updates apollo's store and pushes the same results onto it)

---

@Lepozepo Do you think this is happening more than just one play within Heighliner?

---

It could be happening in other resolvers as the bug was more in the lack of `attributes` from which the query could order from. So I'd look at all the queries using `attributes` and make sure the array contains all used columns. The way the `cache` works right now also seems to cause silent errors if the query fails (which is what made the issue so difficult to identify to begin with). The push I just made is making sure `fetchMoreResolver` does not retrigger/resolve prematurely to protect it against active inefficiencies in render lifecycles, this resolves the root of the duplicate ID's issue (it may have been affecting multiple lists across the app too)